### PR TITLE
CrossCompile: Add a configuration file to cross compile

### PR DESCRIPTION
**Pull request by nokute78**
I add a configuration file.User can cross compile using this file.
### Outline
1. Edit CrossCompile.cmake.
2. cmake -DCMAKE_TOOLCHAIN_FILE=../CrossCompile.cmake ..
3. make

I tried to cross compile for Raspberry Pi2.
(I know http://fluentbit.io/documentation/getting_started/raspberry_pi.html. Just trial.)
- Host: CentOS 6.7 x86_64
- Target: Raspberry pi2(Raspbian)
### Build on Host

Sample CrossCompile.cmake for Raspberry Pi2.  (replace /PATH/TO/ with proper cloned path.)

```
set (CMAKE_SYSTEM_NAME Linux )

set (CMAKE_FIND_ROOT_PATH /PATH/TO/tools/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian )

set (CMAKE_C_COMPILER   ${CMAKE_FIND_ROOT_PATH}/bin/arm-linux-gnueabihf-gcc )
set (CMAKE_CXX_COMPILER ${CMAKE_FIND_ROOT_PATH}/bin/arm-linux-gnueabihf-g++ )
#set (CMAKE_INCLUDE_PATH ${CMAKE_FIND_ROOT_PATH}/include )
#set (CMAKE_LIBRARY_PATH ${CMAKE_FIND_ROOT_PATH}/lib )

set (CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
set (CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
set (CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
```

Note: Some plugins need zlib.h. But, the toolchain doesn't have it.So, I didn't build some plugins.
1. $ git clone https://github.com/raspberrypi/tools.git
2. $ git clone -b cross_compile https://github.com/nokute78/fluent-bit.git
3. $ cd fluent-bit/build
4. Edit CrossCompile.cmake.
5. $ cmake -DCMAKE_TOOLCHAIN_FILE=../CrossCompile.cmake -DWITH_OUT_ES=No -DWITH_OUT_TD=No ..
6. $ make
7. then ARM binary generated.

```
$ file bin/fluent-bit 
bin/fluent-bit: ELF 32-bit LSB executable, ARM, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.26, not stripped
```
### Result on Raspberry Pi

```
pi@raspberrypi /tmp $ ./fluent-bit -i mem -o stdout
Fluent-Bit v0.6.0
Copyright (C) Treasure Data

[2016/01/24 03:34:16] [ info] starting engine
[0] [1453606457, {"total"=>948140, "free"=>899980}]
[1] [1453606458, {"total"=>948140, "free"=>899980}]
[2] [1453606459, {"total"=>948140, "free"=>900012}]
[3] [1453606460, {"total"=>948140, "free"=>900012}]
```

Signed-off-by: nokute78 nokute78@gmail.com

### CODE REVIEW

**Comment by edsiper** [comment]
Great idea, I like it. Let's merge it but there some things that needs to be improved:

- we don't want people have to "learn" how CMake works, I mean tweak a CMakeLists.txt file. Why not just put some variables that when set from the command line can include crosscompile file and set the absolute path to the toolchain ?

**Comment by nokute78** [comment]
Thank you for your advice.  I agree to decrease CMake dependences.
I'll rewrite my patch and pull request.
* Move settings to CMakeLists.txt, to hide from user and to be configurable by options.
* Remove CrossCompile.cmake

But, in this case, options tend to be so long.
(Some absolute paths and toolchain names which often contain platform name.
e.g. arm-linux-gnueabihf-gcc)
This is why I add CrossCompile.cmake.
Another way may be needed.

I think easy cross-compiling makes the project more attractive. 
Thanks!

**Comment by edsiper** [comment]
Sorry, I did not explain clearly:

- I agree that a CrossCompile.cmake it's good, so we handle all details inside that.
- Expose some variable in the __parent__ CMakeLists.txt so CrossCompile.cmake gets included, e.g:

```
$ cmake -DTOOLCHAIN=/path/to ../
```

so 

```
if(TOOLCHAIN)
  include(....)
endif()
``

make sense ?

**Comment by nokute78** [comment]
Sorry for my pointless comment.

I tried to load CrossCompile.cmake in parent CMakeLists.txt , but it didn't work.
CMakeLists.txt may not be able to contain CMAKE_TOOLCHAIN_FILE and such settings.

https://cmake.org/pipermail/cmake/2008-April/021122.html
https://cmake.org/cmake/help/v3.0/variable/CMAKE_TOOLCHAIN_FILE.html#variable:CMAKE_TOOLCHAIN_FILE

**Comment by edsiper** [comment]
what about to have a toolchain file for different use cases ?, as of now I have one for ksz9692 (armadillo-iot), you have one for RPI2, so we can put them under a new cmake.toolchains/, what do you think ?

**Comment by nokute78** [comment]
> what about to have a toolchain file for different use cases ?

Well, we can do 
* put "template" toolchain file under cmake/
* write a document how to use template.

Many parameters depend on user environment.
(Where is a toolchain on host? What is a target arch?  What are libraries which toolchain has? and so on)
So, user must edit our template file.

We can tell user "how to edit it" and "how to use a template file to build" and "trouble shooting".

**Comment by edsiper** [comment]
definitely the user always will need to modify it, which is fine.

File: CrossCompile.cmake [hunk 1]
```diff
@@ -0,0 +1,12 @@
+set (CMAKE_SYSTEM_NAME Linux )
+
+set (CMAKE_FIND_ROOT_PATH /path/to/root )
+
+set (CMAKE_C_COMPILER   ${CMAKE_FIND_ROOT_PATH}/path/to/c_compiler )
```
Comment (cosmo0920) [inline CrossCompile.cmake]: How about using the following?

```cmake
SET(CMAKE_SYSTEM_NAME Linux)
include(CMakeForceCompiler)
IF("${GNU_HOST}" STREQUAL "")
    SET(GNU_HOST arm-poky-linux-gnueabi)
ENDIF()
# Prefix detection only works with compiler id "GNU"
CMAKE_FORCE_C_COMPILER(${GNU_HOST}-gcc GNU)
CMAKE_FORCE_C_COMPILER(${GNU_HOST}-g++ GNU)
```

**Comment by edsiper** [comment]
note this is only a template or guide, not specific for a toolchain.

**Comment by edsiper** [comment]
I'd close this PR since I don't see general problems on people trying to cross compile it.

**Comment by nokute78** [comment]
@edsiper Thanks!
It has been closed at last.

**Pull request closed without merging**